### PR TITLE
Rando-specific display buffer and simplified text printing implementation

### LINKED ISSUE #2257: Custom Music Crash on VC
https://ootrandomizer.com/seed/get?id=1710952
https://clips.twitch.tv/EndearingWrongShrimpSoonerLater-tANKaYkGBRMUt-dc
[OoTR_1710952_G9D4LU1LV2_Cosmetics.json](https://github.com/user-attachments/files/16198108/OoTR_1710952_G9D4LU1LV2_Cosmetics.json)

Custom music on child diving minigame crashed consistently. This was rolled on main dev, seed link and cosmetic file attached.
Custom music was patched on web, played on Wii VC. 

Crashes have happened occasionally on realrob's branch however this was the first consistent crash that was able to be replicated repeatedly.

### PR DESCRIPTION
**Pull request by flagrama**
Adds a unique display buffer for all of the overlays rando creates instead of piggybacking off of the "overlay" display buffer. This gives more control over how much memory is actually being used by randomizer and should reduce random crashes related to overflowing the existing buffer.

This also simplifies the current overlaid text printing implementation. Instead of buffering all of the text and then having to explicitly flush it to the overlay in chunks, text will just be printed directly to the rando display buffer making it easier to add or change the overlaying text in rando.

Also fixes a bug where `sprite_texture` draws a sprite with a huge bottom y position.

### CODE REVIEW

File: ASM/c/gfx.c [hunk 1]
```diff
@@ -164,7 +168,7 @@ void sprite_texture(z64_disp_buf_t* db, sprite_t* sprite, int tile_index, int16_
         G_TX_NOLOD
     );
 
-    gSPTextureRectangle(db->p++, left * 4, top * 4, (left + width) * 4, (top * height) * 4, G_TX_RENDERTILE, 0,0,width_factor, height_factor);
+    gSPTextureRectangle(db->p++, left * 4, top * 4, (left + width) * 4, (top + height) * 4, G_TX_RENDERTILE, 0,0,width_factor, height_factor);
```
Comment (rrealmuto) [inline ASM/c/gfx.c:176]: Recommend making the same fix below in sprite_texture_4b

File: ASM/c/main.c [hunk 2]
```diff
@@ -48,19 +47,30 @@ void before_game_state_update() {
     display_misc_messages();
 }
 
+void close_rando_display_buffer() {
+    OPEN_DISPS(z64_ctxt.gfx);
+
```
Comment (rrealmuto) [inline ASM/c/main.c:54]: recommend adding something to force a crash if rando_db->p exceeds the size. something like

```suggestion
if(((int)rando_db->p - (int)rando_db->buf) > rando_db->size) {
        *((int*)0) = 0;
    }
```

File: ASM/c/gfx.c [hunk 3]
```diff
@@ -214,6 +216,12 @@ void sprite_draw(z64_disp_buf_t* db, sprite_t* sprite, int tile_index,
             width_factor, height_factor);
 }
 
+void rando_display_buffer_init() {
```
Comment (rrealmuto) [inline ASM/c/gfx.c:226]:
```suggestion
#define RANDO_DB_SIZE 0x4000
void rando_display_buffer_init() {
```

File: ASM/c/gfx.c [hunk 4]
```diff
@@ -214,6 +216,12 @@ void sprite_draw(z64_disp_buf_t* db, sprite_t* sprite, int tile_index,
...
+    rando_db = heap_alloc(sizeof(z64_disp_buf_t));
+    rando_db->buf = heap_alloc(0x1000);
```
Comment (rrealmuto) [inline ASM/c/gfx.c]:
```suggestion
    rando_db->size = RANDO_DB_SIZE;
    rando_db->buf = heap_alloc(RANDO_DB_SIZE);
```

**Review by rrealmuto — COMMENTED** [review]
Just a few things

File: ASM/c/main.c (hunk 2 above)
Comment (flagrama) [reply ASM/c/main.c:54]: Even though it won't display on VC, I've got a better idea... (should say rando display buffer though, copy and paste error,  same with the "size" value)

Screenshot 2024-07-15 21-09-10

File: ASM/c/gfx.c (hunk 3 above)
Comment (flagrama) [reply ASM/c/gfx.c:226]: Went with `const unsigned int`s near where the buffer pointer is defined at the top.

File: ASM/c/text.c [hunk 5]
```diff
@@ -2,71 +2,34 @@
 
 #include "gfx.h"
 #include "util.h"
-#include "z64.h"
+#include "z64.h"\
```
Comment (fenhl) [inline ASM/c/text.c]:
```suggestion
#include "z64.h"
```

File: ASM/c/text.c [hunk 6]
```diff
@@ -2,71 +2,34 @@
...
-    char c;
-    int count = 0;
-    while (c = *(s++)) {
-        if (text_end >= text_buf + text_max_chars) break;
-        text_end->c = c;
-        text_end->left = left;
-        text_end->top = top;
-        text_end++;
+int text_print_size(z64_disp_buf_t* db, const char* s, int left, int top, int width, int height) {
+    while(*s != 0x00) {
```
Comment (fenhl) [inline ASM/c/text.c]:
```suggestion
    while (*s != 0x00) {
```

File: ASM/c/triforce.c [hunk 7]
```diff
@@ -90,14 +90,12 @@ void draw_triforce_count(z64_disp_buf_t* db) {
     gDPSetCombineMode(db->p++, G_CC_MODULATEIA_PRIM, G_CC_MODULATEIA_PRIM);
     gDPSetPrimColor(db->p++, 0, 0, 0xDA, 0xD3, 0x0B, alpha);
 
-    text_print(text , draw_x, draw_y_text);
+    text_print(db, text , draw_x, draw_y_text);
```
Comment (fenhl) [inline ASM/c/triforce.c]: This code was already formatted this way but might as well fix it.

```suggestion
    text_print(db, text, draw_x, draw_y_text);
```

File: ASM/c/z64.h [hunk 8]
```diff
@@ -2441,4 +2441,7 @@ typedef void(*z64_Play_SetupRespawnPoint_proc)(z64_game_t *game, int32_t respawn
 #define ITEMGETINF_3B 0x3B
 #define ITEMGETINF_3F 0x3F
 
+extern void Fault_AddHungupAndCrashImpl(const char* msg1, const char* msg2);
+extern int32_t sprintf(char *dst,char *fmt,...);
```
Comment (fenhl) [inline ASM/c/z64.h]:
```suggestion
extern int32_t sprintf(char* dst, char* fmt, ...);
```

**Review by fenhl — CHANGES_REQUESTED** [review]
Just a couple code style nits, seems good otherwise.

**Review by fenhl — APPROVED** [review]

**Pull request merged**